vars/karmaExecuteTests.groovy

```diff
@@ -0,0 +1,73 @@
+    'dockerEnvVars', //envVars to be set in the execution container if required
+    'dockerImage', //Docker image for code execution
+    'dockerName', //name of the Docker container. If not on Kubernetes pod, this will define the network-alias to the NPM container and is thus required for accessing the server, example http://karma:9876 (default).
+    'dockerWorkspace', //user home directory for Docker execution. This will only take effect inside a Kubernetes pod.
+    'failOnError',
+    'installCommand',
+    'modules',
+    'runCommand',
+    'sidecarEnvVars', //envVars to be set in Selenium container if required
+    'sidecarImage', //image for Selenium execution which runs as sidecar to dockerImage
+    'sidecarName', //name of the Selenium container. If not on Kubernetes pod, this will define the network-alias to the Selenium container and is thus required for accessing the server, example http://selenium:4444 (default)
+    'sidecarVolumeBind', //volume bind. This will not take effect in Kubernetes pod.
+    'stashContent' //list of stash names which are required to be unstashed before test run
+]
+@Field Set STEP_CONFIG_KEYS = GENERAL_CONFIG_KEYS
+@Field Set PARAMETER_KEYS = STEP_CONFIG_KEYS
+
+void call(Map parameters = [:]) {
+    handlePipelineStepErrors(stepName: STEP_NAME, stepParameters: parameters) {
+        def script = parameters?.script ?: [commonPipelineEnvironment: commonPipelineEnvironment]
+        def utils = parameters?.juStabUtils ?: new Utils()
+
+        // load default & individual configuration
+        Map config = ConfigurationHelper.newInstance(this)
+            .loadStepDefaults()
+            .mixinGeneralConfig(script.commonPipelineEnvironment, GENERAL_CONFIG_KEYS)
+            .mixinStepConfig(script.commonPipelineEnvironment, STEP_CONFIG_KEYS)
+            .mixinStageConfig(script.commonPipelineEnvironment, parameters.stageName?:env.STAGE_NAME, STEP_CONFIG_KEYS)
+            .mixin(parameters, PARAMETER_KEYS)
+            .use()
+
+        utils.pushToSWA([step: STEP_NAME], config)
+
+        def testJobs = [:]
+        def options = [
+            containerPortMappings: config.containerPortMappings,
+            dockerEnvVars: config.dockerEnvVars,
+            dockerImage: config.dockerImage,
+            dockerName: config.dockerName,
+            dockerWorkspace: config.dockerWorkspace,
+            failOnError: config.failOnError,
+            sidecarEnvVars: config.sidecarEnvVars,
+            sidecarImage: config.sidecarImage,
+            sidecarName: config.sidecarName,
+            sidecarVolumeBind: config.sidecarVolumeBind,
+            stashContent: config.stashContent
+        ]
+        for(String path : config.modules){
+            testJobs["Karma - ${path}"] = {
+                seleniumExecuteTests(options){
+                    sh "cd '${path}' && ${config.installCommand}"
```

Comment:
What is the reason for having two  `sh` statements? What would be wrong with ```cd "${path}" && ${config.installCommand} && ${config.runCommand}```

Not sure if double quotes for surrounding the path would be better. But maybe not to expand variables is a feature ...

Reply:
This way we have a good visual separation in blue ocean an can fold/unfold the commands separately.

Reply:
mmmhhhh ... Ok, that is a valid point for keeping multiple sh statements. I was not well aware about that point, having performance in mind basically.